src/frontend/src/optimizer/plan_node/col_pruning.rs

```diff
@@ -48,3 +52,40 @@ macro_rules! impl_prune_col {
 }
 for_batch_plan_nodes! { impl_prune_col }
 for_stream_plan_nodes! { impl_prune_col }
+
+#[derive(Debug, Clone)]
+pub struct ColumnPruningContext {
+    share_required_cols_map: HashMap<i32, Vec<Vec<usize>>>,
+    share_parent_counter: ShareParentCounter,
+}
+
+impl ColumnPruningContext {
+    pub fn new(root: PlanRef) -> Self {
+        let mut share_parent_counter = ShareParentCounter::default();
+        share_parent_counter.visit(root.clone());
```

Comment:
I used to think this counter should be maintained in the ShareOperator, but in our design, adding a parent on a plan node is not restrained, so maintaining it is so hard. Calculating the counter every time is LGTM.

Reply:
Indeed, I had tried to maintain it manually but finally found that it is too damn hard.